### PR TITLE
Implemented getRemoteAttestation using an async callback function

### PR DESCRIPTION
**Pull request by lfarrel6**

### CODE REVIEW

File: lib/real.js [hunk 1]
```diff
@@ -89,9 +90,68 @@ SecureWorker.getQuoteData = function getQuoteData(quote) {
...
+  const ra_url = "https://api.trustedservices.intel.com/sgx/attestation/v3/report";
+  var payload = {"isvEnclaveQuote":quote_str};
+
+  //populate Ocp-Apim-Subscription-Key with the subscription key from the API portal
+  var options = {
+    method: "POST",
+    headers: {
+      "Ocp-Apim-Subscription-Key": ""
+    }
+  };
```
Comment (shanecurran) [inline lib/real.js]: maybe populate the Ocp-Apim-Subscription-Key from a getRemoteAttestation() parameter

File: lib/real.js [hunk 2]
```diff
@@ -89,9 +90,68 @@ SecureWorker.getQuoteData = function getQuoteData(quote) {
...
-SecureWorker.getRemoteAttestation = function getRemoteAttestation(quote) {
-  throw new Error("Not implemented.");
+// Input is an arraybuffer containing the quote, and a callback function structured at function(result,err)
+// The RA signature is returned using the callback function as an arraybuffer
+// If an error occurs, result will be undefined and err will contain a string
+SecureWorker.getRemoteAttestation = function getRemoteAttestation(quote,callbackFn) {
+  var self=this;
+  var quote_data = self.getQuoteData(quote);
+
+  var quote_str = String.fromCharCode.appply(null,new UInt16Array(quote_data));
```
[the remark this answers is not in the document]
Comment (lfarrel6) [reply lib/real.js]: Fixed in next commit

File: lib/real.js (hunk 1 above)
Comment (lfarrel6) [reply lib/real.js]: Added in next commit

File: lib/real.js [hunk 3]
```diff
@@ -89,9 +90,68 @@ SecureWorker.getQuoteData = function getQuoteData(quote) {
...
+      "Ocp-Apim-Subscription-Key": ""
+    }
+  };
+
+  var handleResponse = function(res){
+    if(res){
+      switch(res.statusCode){
+      case 200:
+        var ra_sig = res.headers["X-IASReport-Signature"];
+        if(ra_sig && ra_sig !== ""){
```
[the remark this answers is not in the document]
Comment (lfarrel6) [reply lib/real.js]: fixed in next commit

**Comment by lfarrel6** [comment]
intel specifies 2 optional parameters on the remote attestation endpoint for a pseManifest and a nonce, should they be included in the function's parameters?

Possible refactor:
```javascript
//preES2015
//payloadObj with structure {subscriptionKey: 'foo', pseManifest: 'bar', nonce: 'baz'}
function getRemoteAttestation(quote, payloadObj, callbackFn) {
...


//post JS update with destructuring
function getRemoteAttestation(quote, {subscriptionKey, pseManifest, nonce}, callbackFn) {
...
```

**Comment by shanecurran** [comment]
Good idea. For now let's use the pre-ES2015 version because the recommended Node version for the package is 6.10.0, which doesn't support restructuring

```javascript
//preES2015
//payloadObj with structure {subscriptionKey: 'foo', pseManifest: 'bar', nonce: 'baz'}
function getRemoteAttestation(quote, payloadObj, callbackFn) { }
```

**Comment by shanecurran** [comment]
Update README to explain the new parameters for getRemoteAttestation()

File: lib/real.js [hunk 4]
```diff
@@ -89,9 +90,79 @@ SecureWorker.getQuoteData = function getQuoteData(quote) {
...
+// Input is an arraybuffer containing the quote, Subscription Key and a callback function structured at function(result,err)
+// The RA signature is returned using the callback function as an arraybuffer
+// If an error occurs, result will be undefined and err will contain a string
+SecureWorker.getRemoteAttestation = function getRemoteAttestation(quote, subscription_key, callbackFn) {
+  var self=this;
+  
+  var quote_data = self.getQuoteData(quote);
+  var quote_str = String.fromCharCode.apply(null,new UInt16Array(quote_data));
+
+  var ra_url = "https://api.trustedservices.intel.com/sgx/attestation/v3/report";
```
Comment (mitar) [inline lib/real.js]: SGX now also allows non-Intel attestation servers, so this could also be an option to the call?

**Comment by mitar** [comment]
Hm, some tests would be useful eventually. :-)

**Comment by shanecurran** [comment]
> Hm, some tests would be useful eventually. :-)

Working on CircleCI integration and unit tests at the moment 👍

File: lib/real.js (hunk 4 above)
Comment (lfarrel6) [reply lib/real.js]: Definitely, I'll add to the next commit

**Comment by mitar** [comment]
> Nearly there! Lint with the [AirBnB style guide](https://github.com/airbnb/javascript) would tidy things up as well 100

I suggest linting should be done in some other PR. It is hard to review things if both code style changes (of unrelated code) and code changes are in the same PR.

File: lib/real.js (hunk 4 above)
Comment (mitar) [reply lib/real.js]: See: https://software.intel.com/en-us/blogs/2018/12/09/an-update-on-3rd-party-attestation

**Pull request merged**